### DISCUSSION TITLE
Support for STK500 devices with STK500_XTAL != 7372800

### DISCUSSION THREAD

**Discussion by Ho-Ro** [avrdudes/avrdude#1560] [Ideas]
The stk500 v1 (as well as the stk500 v2) uses the constant xtal frequency (7372800 Hz) of the original Atmel STK500 HW when calculating the timer parameters `Parm_STK_OSC_PSCALE` and `Parm_STK_OSC_CMATCH` from the terminal command `fosc` to set the external clock for targets without their own clock source. 
These two parameters are direct register values for an 8-bit timer of an ATmega, (e.g. `TCCR2B` and `OCR2A` for m328p).
When using an STK500 compatible programmer with a different clock (e.g. based on an Arduino Nano with 16 MHz), the resulting frequency is much too high. 

My current *solution* is to change this line in `stk500.c` from
```c
#define STK500_XTAL 7372800U
```
to
```c
#define STK500_XTAL 16000000U 
```
before building :)

I don't see a straightforward concept at the moment how to provide the correct frequency for `avrdude`, unless you introduce a new `programmer` parameter, e.g. `xtal`.
I am posting this here for discussion anyway, maybe you can find a simple future solution that fits into your concept.

Thx, Martin

**Comment by mcuee** [comment]
I think it is okay to have `-x xtal` for stk500v1 and stk500v2.

**Reply by mcuee** [reply]
@Ho-Ro 

Maybe you want to create a PR to add xtal to avrdude stk500v1/stk500v2.

**Reply by MCUdude** [reply]
I'd like to hear what @stefanrueger thinks, but I can probably create a PR that allows for `-x xtal`

**Comment by Ho-Ro** [comment]
I already started to deep dive into stk500v1 (`stk500.c`), there are also some minor errors in `-x` handling that I also can fix.
stk500v2 will come next.

**Comment by stefanrueger** [comment]
Using a `-x xtal=...` options sounds like a good solution for this problem for the affected programmers.

**Comment by Ho-Ro** [comment]
https://github.com/avrdudes/avrdude/compare/main...Ho-Ro:avrdude:stk500-x_xtal

This is work in progress, step 1 fixed some minor bugs sleeping in the code (it seems that these extra options are not heavily used).

Unfortunately my editor kate automatically removes trailing spaces even when told to do not, so there are some format changes. If you don't like this, I can rework - finally I found [the solution](https://kde-bugs-dist.kde.narkive.com/GAv68boQ/kate-bug-397956-new-kate-unconditionally-removes-trailing-spaces-if-kateconfig-file-doesn-t-exist).

Step 2 will be the implementation of `-x xtal=nnn` for `stk500.c` before I go on with v2.

EDIT: HOLD on, the more I test, the more bugs creep in, e.g. -xfosc=100k works, but -xfosc=100000 sets fosc=0.000
I will fix this together with step2 - just pushed to my repo.
Good night!

**Reply by MCUdude** [reply]
Great start! We probably add support for xtal "units" as well, just like we can with `fosc` (you can most likely just copy-paste the code).

```
-x xtal=16000000
-x xtal=16MHz (or Mhz)
-x xtal=16M
-x xtal=1600k
-x xtal=1600kHz (or khz)
```

**Comment by Ho-Ro** [comment]
>  (you can most likely just copy-paste the code).

I did - and during testing I discovered some *issues* how we call it in the job - not bug, not error...

> -x xtal=16MHz (or Mhz)

or even mhz (shudder)

but -x xtal=1234567Hz should also work, I will add this. - *wenn schon, denn schon* - how we say in German. :)

**Comment by mcuee** [comment]
* #1566 has been merged.
